### ISSUE TITLE
Field filters break SQL-type questions for Redshift data sources in v0.36.0

### ISSUE THREAD

**Issue by cgddrd** [metabase/metabase#12984]
**Describe the bug**

As of the v0.36.0 release, all types of dynamic field filter variables break SQL-type questions in Metabase. From our testing, this appears to only be affecting a subset of data source types - in our case, **Redshift**. This problem **does not** appear to be happening for the Sample Dataset (H2) or MySQL connections. When a field filter is applied, all questions give the same error message (regardless of schema/table in our Redshift database): 

![metabase-bug](https://user-images.githubusercontent.com/2536819/88278266-01999280-ccda-11ea-9d11-070592ccd5f4.png)

Weirdly, if you **remove the field filter on the top bar of the editor UI** and instead set/rely upon a **default value** for the field filter, everything works as expected and the query returns results. Therefore from our experience, it seems overriding the default value for a field filter causes the issue, but we've not been able to confirm that this is the only trigger event. 

Also, this issue **does not occur** with questions built with the **GUI editor**, it only seems to be affecting custom SQL-type questions (of which we have many).

We've attempted to fix the issue by re-syncing the schema via the `Admin -> Databases` tooling, but this made no difference.

**Logs**

```
[45fcda48-7cf1-4c35-bd83-013ace0bb434] 2020-07-23T11:59:39+01:00 ERROR metabase.query-processor.middleware.catch-exceptions Error processing query: null
{:database_id 8,
:started_at #t "2020-07-23T10:59:39.300851Z[GMT]",
:error_type :invalid-query,
:json_query
{:database 8,
:native
{:template-tags
{:date {:id "db8336eb-388e-c193-c8a8-d8cf63d083a3", :name "date", :display-name "Date", :type "dimension", :dimension ["field-id" 48920], :widget-type "date/all-options", :default nil}},
:query "select *\nfrom \nanalytics.mytable\nwhere\n{{date}}"},
:type "native",
:parameters [{:type "date/all-options", :target ["dimension" ["template-tag" "date"]], :value "past30days"}],
:middleware {:add-default-userland-constraints? true}},
:status :failed,
:class clojure.lang.ExceptionInfo,
:stacktrace
["--> mbql.util$fn__25465$field_clause__GT_id_or_literal__25470.invoke(util.clj:441)"
"driver.redshift$query__GT_field_values$fn__510$match_511__512.invoke(redshift.clj:188)"
"mbql.util.match$match_in_collection$iter__24845__24849$fn__24850.invoke(match.clj:93)"
"mbql.util.match$match_in_collection.invokeStatic(match.clj:92)"
"mbql.util.match$match_in_collection.invoke(match.clj:86)"
"driver.redshift$query__GT_field_values$fn__510$match_511__512.invoke(redshift.clj:188)"
"driver.redshift$query__GT_field_values$fn__510.invoke(redshift.clj:188)"
"driver.redshift$query__GT_field_values.invokeStatic(redshift.clj:181)"
"driver.redshift$query__GT_field_values.invoke(redshift.clj:178)"
"driver.redshift$eval521$fn__523.invoke(redshift.clj:204)"
"driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:382)"
"driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
"driver.sql_jdbc$fn__73809.invokeStatic(sql_jdbc.clj:49)"
"driver.sql_jdbc$fn__73809.invoke(sql_jdbc.clj:47)"
"query_processor.context$executef.invokeStatic(context.clj:59)"
"query_processor.context$executef.invoke(context.clj:48)"
"query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
"query_processor.context.default$default_runf.invoke(default.clj:67)"
"query_processor.context$runf.invokeStatic(context.clj:45)"
"query_processor.context$runf.invoke(context.clj:39)"
"query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
"query_processor.reducible$pivot.invoke(reducible.clj:31)"
"query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45519.invoke(mbql_to_native.clj:26)"
"query_processor.middleware.check_features$check_features$fn__44821.invoke(check_features.clj:42)"
"query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__45684.invoke(optimize_datetime_filters.clj:133)"
"query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47212.invoke(wrap_value_literals.clj:137)"
"query_processor.middleware.annotate$add_column_info$fn__43442.invoke(annotate.clj:574)"
"query_processor.middleware.permissions$check_query_permissions$fn__44696.invoke(permissions.clj:64)"
"query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46202.invoke(pre_alias_aggregations.clj:40)"
"query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__44894.invoke(cumulative_aggregations.clj:61)"
"query_processor.middleware.resolve_joins$resolve_joins$fn__46734.invoke(resolve_joins.clj:183)"
"query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39172.invoke(add_implicit_joins.clj:245)"
"query_processor.middleware.limit$limit$fn__45505.invoke(limit.clj:38)"
"query_processor.middleware.format_rows$format_rows$fn__45486.invoke(format_rows.clj:81)"
"query_processor.middleware.desugar$desugar$fn__44960.invoke(desugar.clj:22)"
"query_processor.middleware.binning$update_binning_strategy$fn__43986.invoke(binning.clj:229)"
"query_processor.middleware.resolve_fields$resolve_fields$fn__44502.invoke(resolve_fields.clj:24)"
"query_processor.middleware.add_dimension_projections$add_remapping$fn__38721.invoke(add_dimension_projections.clj:270)"
"query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__38928.invoke(add_implicit_clauses.clj:141)"
"query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39321.invoke(add_source_metadata.clj:105)"
"query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46399.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
"query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43627.invoke(auto_bucket_datetimes.clj:125)"
"query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44549.invoke(resolve_source_table.clj:46)"
"query_processor.middleware.parameters$substitute_parameters$fn__46184.invoke(parameters.clj:114)"
"query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44601.invoke(resolve_referenced.clj:80)"
"query_processor.middleware.expand_macros$expand_macros$fn__45216.invoke(expand_macros.clj:158)"
"query_processor.middleware.add_timezone_info$add_timezone_info$fn__39352.invoke(add_timezone_info.clj:15)"
"query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47096.invoke(splice_params_in_response.clj:32)"
"query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46410$fn__46414.invoke(resolve_database_and_driver.clj:33)"
"driver$do_with_driver.invokeStatic(driver.clj:61)"
"driver$do_with_driver.invoke(driver.clj:57)"
"query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46410.invoke(resolve_database_and_driver.clj:27)"
"query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45434.invoke(fetch_source_query.clj:267)"
"query_processor.middleware.store$initialize_store$fn__47105$fn__47106.invoke(store.clj:11)"
"query_processor.store$do_with_store.invokeStatic(store.clj:46)"
"query_processor.store$do_with_store.invoke(store.clj:40)"
"query_processor.middleware.store$initialize_store$fn__47105.invoke(store.clj:10)"
"query_processor.middleware.cache$maybe_return_cached_results$fn__44478.invoke(cache.clj:209)"
"query_processor.middleware.validate$validate_query$fn__47114.invoke(validate.clj:10)"
"query_processor.middleware.normalize_query$normalize$fn__45532.invoke(normalize_query.clj:22)"
"query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39190.invoke(add_rows_truncated.clj:36)"
"query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47081.invoke(results_metadata.clj:147)"
"query_processor.middleware.constraints$add_default_userland_constraints$fn__44837.invoke(constraints.clj:42)"
"query_processor.middleware.process_userland_query$process_userland_query$fn__46273.invoke(process_userland_query.clj:136)"
"query_processor.middleware.catch_exceptions$catch_exceptions$fn__44780.invoke(catch_exceptions.clj:174)"
"query_processor.reducible$async_qp$qp_STAR___38004$thunk__38005.invoke(reducible.clj:101)"
"query_processor.reducible$async_qp$qp_STAR___38004.invoke(reducible.clj:107)"
"query_processor.reducible$sync_qp$qp_STAR___38013$fn__38016.invoke(reducible.clj:133)"
"query_processor.reducible$sync_qp$qp_STAR___38013.invoke(reducible.clj:132)"
"query_processor$process_userland_query.invokeStatic(query_processor.clj:213)"
"query_processor$process_userland_query.doInvoke(query_processor.clj:209)"
"query_processor$fn__47256$process_query_and_save_execution_BANG___47265$fn__47268.invoke(query_processor.clj:225)"
"query_processor$fn__47256$process_query_and_save_execution_BANG___47265.invoke(query_processor.clj:217)"
"query_processor$fn__47300$process_query_and_save_with_max_results_constraints_BANG___47309$fn__47312.invoke(query_processor.clj:237)"
"query_processor$fn__47300$process_query_and_save_with_max_results_constraints_BANG___47309.invoke(query_processor.clj:230)"
"api.dataset$fn__50591$fn__50594.invoke(dataset.clj:54)"
"query_processor.streaming$streaming_response_STAR_$fn__35426$fn__35427.invoke(streaming.clj:73)"
"query_processor.streaming$streaming_response_STAR_$fn__35426.invoke(streaming.clj:72)"
"async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:66)"
"async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:64)"
"async.streaming_response$do_f_async$fn__23250.invoke(streaming_response.clj:85)"],
:context :ad-hoc,
:error
"Input to field-clause->id-or-literal does not match schema: \n\n\t [(named (named (not (some-matching-condition? a-clojure.lang.PersistentVector)) \"Must be a valid instance of one of these clauses: :field-id, :field-literal, :joined-field, :fk->, :datetime-field, :expression, :binning-strategy\") clause)] \n\n",
:row_count 0,
:running_time 0,
:ex-data
{:type :schema.core/error,
:value [[:template-tag "date"]],
:error
[(named (named (not (some-matching-condition? a-clojure.lang.PersistentVector)) "Must be a valid instance of one of these clauses: :field-id, :field-literal, :joined-field, :fk->, :datetime-field, :expression, :binning-strategy") clause)]},
:data {:rows [], :cols []}}
```

**To Reproduce**

Use any field filter with a SQL-type question (whether directly on a question or via a dashboard filter) querying a Redshift data source.

**Expected behavior**

Field filters for all types of questions and database sources should filter appropriately and return relevant results. 

**Screenshots**

See above.

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.7+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.7",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.7+10",
    "os.name": "Linux",
    "os.version": "4.14.181-108.257.amzn1.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "redshift",
      "mysql",
      "h2"
    ],
    "hosting-env": "elastic-beanstalk",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.5"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-07-21",
      "tag": "v0.36.0",
      "branch": "release-0.36.x",
      "hash": "b31049f"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**

**Severe** - this issue has broken a large number of production dashboards and source questions, which have required us to remove all field filters and use alternative non-dynamic filtering through bespoke `WHERE` clauses for each affected question.
 
Also, rolling back to v0.35.x **isn't an option**, as breaking changes with the Metabase DB schema cause schema synchronisation and field fingerprinting to fail, which is an even greater problem than this. 

**Additional context**

**Comment by sbelak** [comment]
I can repo this on stats with a `select * from public.checkins where {{date}}` and setting the variable and filter as is in the screenshot.

**Comment by flamber** [comment]
Fixed in #13015 and will be part of 0.36.1, which should be released within a few days.

**Comment by cgddrd** [comment]
Thanks for this swift fix everyone!
